### PR TITLE
[cmake] add liboneflow_cpp target

### PR DESCRIPTION
**Pull request by daquexian**
新增 BUILD_CPP_API option，将 liboneflow_cpp.so 作为包含 c++ api 的 .so，支持同时编译 python .so 和 c++ .so

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -201,19 +202,21 @@ endif()
...
-  set(ONEFLOW_INCLUDE_DIR "${PROJECT_BINARY_DIR}/liboneflow/include/oneflow")
-  set(ONEFLOW_LIBRARY_DIR "${PROJECT_BINARY_DIR}/liboneflow/lib")
-  set(ONEFLOW_SHARE_DIR "${PROJECT_BINARY_DIR}/liboneflow/share")
-  make_directory(${ONEFLOW_INCLUDE_DIR})
-  make_directory(${ONEFLOW_LIBRARY_DIR})
-  make_directory(${ONEFLOW_SHARE_DIR})
+endif(BUILD_PYTHON)
+
+if(BUILD_CPP_API)
+  set(LIBONEFLOW_INCLUDE_DIR "${PROJECT_BINARY_DIR}/liboneflow_cpp/include/oneflow")
```
Comment (jackalcooper) [inline CMakeLists.txt:208]: 这几个目录为什么要放在binary dir，直接存在源码里面然后 gitkeep 可以吗

Comment (daquexian) [reply CMakeLists.txt:208]: 这个感觉应该改成在 make install 时复制头文件，在 cmake 和 build 阶段不复制。这个 PR 先这样吧后续再单独修改。

File: cmake/oneflow.cmake [hunk 2]
```diff
@@ -356,13 +346,28 @@ if(BUILD_PYTHON)
...
+endif()
+
 file(RELATIVE_PATH PROJECT_BINARY_DIR_RELATIVE ${PROJECT_SOURCE_DIR} ${PROJECT_BINARY_DIR})
 
 # build test
 if(BUILD_TESTING)
   if (of_all_test_cc)
     oneflow_add_executable(oneflow_testexe ${of_all_test_cc})
-    target_link_libraries(oneflow_testexe ${of_libs} ${oneflow_third_party_libs} ${oneflow_exe_third_party_libs})
+    target_link_libraries(oneflow_testexe ${oneflow_test_libs})
```
Comment (jackalcooper) [inline cmake/oneflow.cmake:365]: 这个是不是可以考虑为 oneflow cpp 单独创建个测试 target

File: cmake/oneflow.cmake [hunk 3]
```diff
@@ -356,13 +346,28 @@ if(BUILD_PYTHON)
 
 endif(BUILD_PYTHON)
 
+if (BUILD_CPP_API)
+  if(BUILD_SHARED_LIBONEFLOW)
+    oneflow_add_library(oneflow_cpp SHARED ${of_cpp_api_obj_cc})
+  else()
+    oneflow_add_library(oneflow_cpp ${of_cpp_api_obj_cc})
+  endif()
+  set_target_properties(oneflow_cpp PROPERTIES ARCHIVE_OUTPUT_DIRECTORY "${LIBONEFLOW_LIBRARY_DIR}" LIBRARY_OUTPUT_DIRECTORY "${LIBONEFLOW_LIBRARY_DIR}")
```
Comment (jackalcooper) [inline cmake/oneflow.cmake:355]: 这种属性可以统一用一个 install 来写吗

File: cmake/oneflow.cmake [hunk 4]
```diff
@@ -119,19 +119,18 @@ foreach(oneflow_single_file ${oneflow_all_src})
...
 
-  else() # build_python
-
+  if(BUILD_CPP_API)
     if("${oneflow_single_file}" MATCHES "^${PROJECT_SOURCE_DIR}/oneflow/api/cpp/.*\\.(h|cpp)$")
       if("${oneflow_single_file}" MATCHES "^${PROJECT_SOURCE_DIR}/oneflow/api/cpp/.*_test\\.cpp$")
         list(APPEND of_all_test_cc ${oneflow_single_file})
       else()
-        list(APPEND of_all_obj_cc ${oneflow_single_file})
+        list(APPEND of_cpp_api_obj_cc ${oneflow_single_file})
```
Comment (jackalcooper) [inline cmake/oneflow.cmake:129]: 可以考虑把cpp_api的测试单独写个target？然后不走这里的glob，直接显式列文件

Comment (daquexian) [reply cmake/oneflow.cmake:129]: 可以的

Comment (daquexian) [reply cmake/oneflow.cmake:129]: 显式列文件感觉不太科学，我改成为 cpp api 单独写一个 glob 吧

File: cmake/oneflow.cmake (hunk 3 above)
Comment (daquexian) [reply cmake/oneflow.cmake:355]: 是可以的，下一个 PR 一起解决吧

**Review by jackalcooper — APPROVED** [review]

**Review by mosout — APPROVED** [review]

**Comment by ouyangyu** [comment]
<p align="center">表 2021-12-12 ResNet50 lazy  dlperf 运行结果</p>

|                                         | [56152a2](https://github.com/Oneflow-Inc/oneflow/commit/56152a2) |
| :-------------------------------------: | :----------------------------------------------------------: |
|                                         |                    Memory/Throughput                         |
| ResNet50_lazy_dlperf_gpudecode_FP16_b512_1n1g | 23933 MiB / [1416.45](https://oneflow-test.oss-cn-beijing.aliyuncs.com/OneBrain/commit/liboneflow_cpp/20211212/56152a2/ResNet50-lazy/dlperf/1n1g/resnet50_lazy_FP16_b512_1n1g_liboneflow_cpp_20211212_223708445429154.log) |
| ResNet50_lazy_dlperf_cpudecode_FP16_b512_1n1g | 23165 MiB / [1433.494](https://oneflow-test.oss-cn-beijing.aliyuncs.com/OneBrain/commit/liboneflow_cpp/20211212/56152a2/ResNet50-lazy/dlperf/1n1g/resnet50_lazy_FP16_b512_1n1g_liboneflow_cpp_20211212_224323128806901.log) |
| ResNet50_lazy_dlperf_gpudecode_FP32_b256_1n1g | 23475 MiB / [388.256](https://oneflow-test.oss-cn-beijing.aliyuncs.com/OneBrain/commit/liboneflow_cpp/20211212/56152a2/ResNet50-lazy/dlperf/1n1g/resnet50_lazy_FP32_b256_1n1g_liboneflow_cpp_20211212_223943450894648.log) |
| ResNet50_lazy_dlperf_cpudecode_FP32_b256_1n1g | 22795 MiB / [389.344](https://oneflow-test.oss-cn-beijing.aliyuncs.com/OneBrain/commit/liboneflow_cpp/20211212/56152a2/ResNet50-lazy/dlperf/1n1g/resnet50_lazy_FP32_b256_1n1g_liboneflow_cpp_20211212_224557951696297.log) |



<p align="center">表 2021-12-12 ResNet50 lazy  dlperf 运行结果</p>

|                                         | [56152a2](https://github.com/Oneflow-Inc/oneflow/commit/56152a2) |
| :-------------------------------------: | :----------------------------------------------------------: |
|                                         |                    Memory/Throughput                         |
| ResNet50_lazy_dlperf_gpudecode_FP16_b512_1n8g | 24549 MiB / [10699.531](https://oneflow-test.oss-cn-beijing.aliyuncs.com/OneBrain/commit/liboneflow_cpp/20211212/56152a2/ResNet50-lazy/dlperf/1n8g/resnet50_lazy_FP16_b512_1n8g_liboneflow_cpp_20211212_215921275497623.log) |
| ResNet50_lazy_dlperf_cpudecode_FP16_b512_1n8g | 23771 MiB / [10886.34](https://oneflow-test.oss-cn-beijing.aliyuncs.com/OneBrain/commit/liboneflow_cpp/20211212/56152a2/ResNet50-lazy/dlperf/1n8g/resnet50_lazy_FP16_b512_1n8g_liboneflow_cpp_20211212_220704352320076.log) |
| ResNet50_lazy_dlperf_gpudecode_FP32_b256_1n8g | 24037 MiB / [2978.023](https://oneflow-test.oss-cn-beijing.aliyuncs.com/OneBrain/commit/liboneflow_cpp/20211212/56152a2/ResNet50-lazy/dlperf/1n8g/resnet50_lazy_FP32_b256_1n8g_liboneflow_cpp_20211212_220240345472582.log) |
| ResNet50_lazy_dlperf_cpudecode_FP32_b256_1n8g | 23369 MiB / [3044.615](https://oneflow-test.oss-cn-beijing.aliyuncs.com/OneBrain/commit/liboneflow_cpp/20211212/56152a2/ResNet50-lazy/dlperf/1n8g/resnet50_lazy_FP32_b256_1n8g_liboneflow_cpp_20211212_221019859360696.log) |


<p align="center">表 2021-12-12 ResNet50 graph  dlperf 运行结果</p>

|                                         | [56152a2](https://github.com/Oneflow-Inc/oneflow/commit/56152a2) |
| :-------------------------------------: | :----------------------------------------------------------: |
|                                         |                    Memory/Throughput                         |
| ResNet50_graph_dlperf_cpudecode_FP16_b512_1n1g | 26314 MiB / [660.62](https://oneflow-test.oss-cn-beijing.aliyuncs.com/OneBrain/commit/liboneflow_cpp/20211212/56152a2/ResNet50-graph/dlperf/1n1g/resnet50_graph_DCcpu_FP16_b512_1n1g_liboneflow_cpp_20211212_222428259851738.log) |
| ResNet50_graph_dlperf_cpudecode_FP32_b256_1n1g | 23020 MiB / [361.22](https://oneflow-test.oss-cn-beijing.aliyuncs.com/OneBrain/commit/liboneflow_cpp/20211212/56152a2/ResNet50-graph/dlperf/1n1g/resnet50_graph_DCcpu_FP32_b256_1n1g_liboneflow_cpp_20211212_222832929113581.log) |
| ResNet50_graph_dlperf_gpudecode_FP16_b512_1n1g | 27520 MiB / [773.98](https://oneflow-test.oss-cn-beijing.aliyuncs.com/OneBrain/commit/liboneflow_cpp/20211212/56152a2/ResNet50-graph/dlperf/1n1g/resnet50_graph_DCgpu_FP16_b512_1n1g_liboneflow_cpp_20211212_221710761077115.log) |
| ResNet50_graph_dlperf_gpudecode_FP32_b256_1n1g | 24104 MiB / [360.69](https://oneflow-test.oss-cn-beijing.aliyuncs.com/OneBrain/commit/liboneflow_cpp/20211212/56152a2/ResNet50-graph/dlperf/1n1g/resnet50_graph_DCgpu_FP32_b256_1n1g_liboneflow_cpp_20211212_222043005418272.log) |



<p align="center">表 2021-12-12 ResNet50 graph  dlperf 运行结果</p>

|                                         | [56152a2](https://github.com/Oneflow-Inc/oneflow/commit/56152a2) |
| :-------------------------------------: | :----------------------------------------------------------: |
|                                         |                    Memory/Throughput                         |
| ResNet50_graph_dlperf_cpudecode_FP16_b512_1n8g | 27410 MiB / [5000.88](https://oneflow-test.oss-cn-beijing.aliyuncs.com/OneBrain/commit/liboneflow_cpp/20211212/56152a2/ResNet50-graph/dlperf/1n8g/resnet50_graph_DCcpu_FP16_b512_1n8g_liboneflow_cpp_20211212_214402640463664.log) |
| ResNet50_graph_dlperf_cpudecode_FP32_b256_1n8g | 24008 MiB / [2784.08](https://oneflow-test.oss-cn-beijing.aliyuncs.com/OneBrain/commit/liboneflow_cpp/20211212/56152a2/ResNet50-graph/dlperf/1n8g/resnet50_graph_DCcpu_FP32_b256_1n8g_liboneflow_cpp_20211212_214857746802306.log) |
| ResNet50_graph_dlperf_gpudecode_FP16_b512_1n8g | 28496 MiB / [5759.36](https://oneflow-test.oss-cn-beijing.aliyuncs.com/OneBrain/commit/liboneflow_cpp/20211212/56152a2/ResNet50-graph/dlperf/1n8g/resnet50_graph_DCgpu_FP16_b512_1n8g_liboneflow_cpp_20211212_213513817337703.log) |
| ResNet50_graph_dlperf_gpudecode_FP32_b256_1n8g | 25114 MiB / [2812.56](https://oneflow-test.oss-cn-beijing.aliyuncs.com/OneBrain/commit/liboneflow_cpp/20211212/56152a2/ResNet50-graph/dlperf/1n8g/resnet50_graph_DCgpu_FP32_b256_1n8g_liboneflow_cpp_20211212_213937791892106.log) |
 


<p align="center">表 2021-12-12 ResNet50 ddp  dlperf 运行结果</p>

|                                         | [56152a2](https://github.com/Oneflow-Inc/oneflow/commit/56152a2) |
| :-------------------------------------: | :----------------------------------------------------------: |
|                                         |                    Memory/Throughput                         |
| ResNet50_ddp_dlperf_cpudecode_FP32_b256_1n1g | 27232 MiB / [272.63](https://oneflow-test.oss-cn-beijing.aliyuncs.com/OneBrain/commit/liboneflow_cpp/20211212/56152a2/ResNet50-ddp/dlperf/1n1g/resnet50_ddp_DCcpu_FP32_b256_1n1g_liboneflow_cpp_20211212_223217641200030.log) |



<p align="center">表 2021-12-12 ResNet50 ddp  dlperf 运行结果</p>

|                                         | [56152a2](https://github.com/Oneflow-Inc/oneflow/commit/56152a2) |
| :-------------------------------------: | :----------------------------------------------------------: |
|                                         |                    Memory/Throughput                         |
| ResNet50_ddp_dlperf_cpudecode_FP32_b256_1n8g | 27664 MiB / [1848.48](https://oneflow-test.oss-cn-beijing.aliyuncs.com/OneBrain/commit/liboneflow_cpp/20211212/56152a2/ResNet50-ddp/dlperf/1n8g/resnet50_ddp_DCcpu_FP32_b256_1n8g_liboneflow_cpp_20211212_215324076262870.log) |

ResNet50 性能与历史对比没有问题

**Pull request merged**